iceoryx2-cal/src/zero_copy_connection/used_chunk_list.rs

```diff
@@ -0,0 +1,205 @@
+};
+
+use iceoryx2_bb_elementary::{
+    math::align_to,
+    owning_pointer::OwningPointer,
+    relocatable_container::RelocatableContainer,
+    relocatable_ptr::{PointerTrait, RelocatablePointer},
+};
+use iceoryx2_bb_log::{fail, fatal_panic};
+
+use crate::shm_allocator::PointerOffset;
+
+pub type UsedChunkList = details::UsedChunkList<OwningPointer<AtomicBool>>;
+pub type RelocatableUsedChunkList = details::UsedChunkList<RelocatablePointer<AtomicBool>>;
+
+pub mod details {
+    use std::fmt::Debug;
+
+    use iceoryx2_bb_elementary::owning_pointer::OwningPointer;
+
+    use super::*;
+
+    #[derive(Debug)]
+    #[repr(C)]
+    pub struct UsedChunkList<PointerType: PointerTrait<AtomicBool>> {
+        data_ptr: PointerType,
+        capacity: usize,
+        is_memory_initialized: AtomicBool,
+    }
+
+    unsafe impl<PointerType: PointerTrait<AtomicBool>> Send for UsedChunkList<PointerType> {}
+    unsafe impl<PointerType: PointerTrait<AtomicBool>> Sync for UsedChunkList<PointerType> {}
+
+    impl UsedChunkList<OwningPointer<AtomicBool>> {
+        pub fn new(capacity: usize) -> Self {
+            let mut data_ptr = OwningPointer::<AtomicBool>::new_with_alloc(capacity);
+
+            for i in 0..capacity {
+                unsafe { data_ptr.as_mut_ptr().add(i).write(AtomicBool::new(false)) };
+            }
+
+            Self {
+                data_ptr,
+                capacity,
+                is_memory_initialized: AtomicBool::new(true),
+            }
+        }
+    }
+
+    impl RelocatableContainer for UsedChunkList<RelocatablePointer<AtomicBool>> {
+        unsafe fn new_uninit(capacity: usize) -> Self {
+            Self {
+                data_ptr: RelocatablePointer::new_uninit(),
+                capacity,
+                is_memory_initialized: AtomicBool::new(false),
+            }
+        }
+
+        unsafe fn init<T: iceoryx2_bb_elementary::allocator::BaseAllocator>(
+            &self,
+            allocator: &T,
+        ) -> Result<(), iceoryx2_bb_elementary::allocator::AllocationError> {
+            if self.is_memory_initialized.load(Ordering::Relaxed) {
+                fatal_panic!(from self,
+                "Memory already initialized. Initializing it twice may lead to undefined behavior.");
+            }
+
+            let memory = fail!(from self, when allocator
+            .allocate(Layout::from_size_align_unchecked(
+                    std::mem::size_of::<AtomicBool>() * self.capacity,
+                    std::mem::align_of::<AtomicBool>())),
+            "Failed to initialize since the allocation of the data memory failed.");
+
+            self.data_ptr.init(memory);
+
+            for i in 0..self.capacity {
+                unsafe {
+                    (self.data_ptr.as_ptr() as *mut AtomicBool)
+                        .add(i)
+                        .write(AtomicBool::new(false))
+                };
+            }
+
+            self.is_memory_initialized.store(true, Ordering::Relaxed);
```

Comment:
Can you document why relaxed is here sufficient? Is there a way the member could be read from another thread?

Reply:
Nope. The RelocatableContainer is not allowed to be used, send, synced, shared with any other thread when it is not initialized - meaning `init` was not yet called.

The RelocatableContainer has a two-step initialization approach
1. Call `new_uninit`
2. Call `init`

instead of a simple `new` call. Therefore, there is no need for syncing since the container is at this point in time only accessible from one thread.